### PR TITLE
Updated MT branch

### PR DESCRIPTION
**Pull request by Kelina**
Introducing a recurrent sequence-to-sequence model for MT.

### CODE REVIEW

**Comment by sleepinyourhat** [comment]
Thanks! What results/sanity checks do you have so far?

**Comment by sleepinyourhat** [comment]
And more generally, what's the state of MT/seq2seq support right now? What's verified to work well, what should work, and are there any possible configurations to avoid?

**Comment by Kelina** [comment]
> Thanks! What results/sanity checks do you have so far?

I trained one model with similar hyperparameters to [https://arxiv.org/pdf/1508.04025.pdf](https://arxiv.org/pdf/1508.04025.pdf). The new model obtained 21.12 BLEU on the validation set, so it's roughly comparable.

**Comment by Kelina** [comment]
> And more generally, what's the state of MT/seq2seq support right now? What's verified to work well, what should work, and are there any possible configurations to avoid?

For now, morphological segmentation as in the respective config file works. MT works as in the respective config file as well; no other configurations have been tried.

**Comment by sleepinyourhat** [comment]
Could you say more about the exact comparison you're making for WMT. Is it test vs. test? What's the target number you're trying to match?

File: jiant/config/wmt14_en_de.conf [hunk 1]
```diff
@@ -0,0 +1,51 @@
+// Template for translation as a seq2seq problem.
```
Comment (sleepinyourhat) [inline jiant/config/wmt14_en_de.conf]: Say a bit more about where these values come from. Are they supposed to match a paper? Which ones are necessary parts of the setup, and which are just things that you tuned?

Comment (sleepinyourhat) [reply jiant/config/wmt14_en_de.conf]: Mention that this is a non-Transformers setup, and say what it'd take to use a Transformers model.

File: jiant/tasks/seq2seq.py [hunk 2]
```diff
@@ -22,14 +23,18 @@
...
 
 @register_task("seg_wix", rel_path="seg/wix/", max_targ_v_size=200)
+@register_task("wmt14_en_de", rel_path="wmt14/en_de/", max_targ_v_size=40000)
 class Seq2SeqTask(SequenceGenerationTask):
     """Sequence-to-sequence Task"""
 
     def __init__(self, path, max_seq_len, max_targ_v_size, name, **kw):
         super().__init__(name, **kw)
-        self.scorer2 = BooleanAccuracy()
+        if name == "seg_wix":
```
Comment (sleepinyourhat) [inline jiant/tasks/seq2seq.py]: Cleaner to make this a keyword argument (like max_targ_v_size) rather than a switch on name.

File: jiant/tasks/seq2seq.py [hunk 3]
```diff
@@ -71,7 +72,10 @@ def get_data_iter(self, path):
                 if len(row) < 2 or not row[0] or not row[1]:
                     continue
                 src_sent = tokenize_and_truncate(self._tokenizer_name, row[0], self.max_seq_len)
-                tgt_sent = tokenize_and_truncate(self._tokenizer_name, row[2], self.max_seq_len)
+                if self.name == "seg_wix":
```
Comment (sleepinyourhat) [inline jiant/tasks/seq2seq.py]: Use a specific property of the task object, rather than the name. (It's fairly common to override name if you want to do MTL on multiple datasets for the same task.)

Comment (sleepinyourhat) [reply jiant/tasks/seq2seq.py]: Same for similar if statements below.

File: jiant/tasks/seq2seq.py [hunk 4]
```diff
@@ -128,10 +137,33 @@ def update_metrics(self, logits, labels, tagmask=None, predictions=None):
             # Cut labels if predictions (without gold target) are shorter.
             labels = labels[:, : predictions.shape[1]]
             tagmask = tagmask[:, : predictions.shape[1]]
```
Comment (sleepinyourhat) [inline jiant/tasks/seq2seq.py]: Are we now creating this even in cases where we won't use it?

File: jiant/trainer.py [hunk 5]
```diff
@@ -860,7 +860,8 @@ def _calculate_validation_performance(
 
             all_val_metrics["%s_loss" % task.name] += loss.data.cpu().numpy()
             n_examples += get_output_attribute(out, "n_exs", self._cuda_device)
-            if print_output:
+            # The following doesn't work for other inputs, yet.
+            if print_output and "inputs" in batch and "words" in batch["inputs"]:
```
Comment (sleepinyourhat) [inline jiant/trainer.py]: This seems like a place where an assert would help: Right now, won't this cause jiant to just ignore a config option?

File: scripts/wmt14/prepare_MT_data.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import sys
```
Comment (sleepinyourhat) [inline scripts/wmt14/prepare_MT_data.py]: Use lowercase for filenames, as elsewhere.

File: scripts/wmt14/README [hunk 7]
```diff
@@ -0,0 +1,20 @@
+For machine translation (MT) tasks, all examples should be in the format
```
Comment (sleepinyourhat) [inline scripts/wmt14/README]: Say where the example data comes from.

Comment (Kelina) [reply scripts/wmt14/README]: Done.

File: scripts/wmt14/prepare_MT_data.py (hunk 6 above)
Comment (Kelina) [reply scripts/wmt14/prepare_MT_data.py]: Done.

File: jiant/trainer.py (hunk 5 above)
Comment (Kelina) [reply jiant/trainer.py]: I made it a NotImplementedError. Does that seem okay?

File: jiant/config/wmt14_en_de.conf (hunk 1 above)
Comment (Kelina) [reply jiant/config/wmt14_en_de.conf]: Done.

File: jiant/tasks/seq2seq.py (hunk 2 above)
Comment (Kelina) [reply jiant/tasks/seq2seq.py]: Done.

File: jiant/tasks/seq2seq.py (hunk 3 above)
Comment (Kelina) [reply jiant/tasks/seq2seq.py]: Done.

File: jiant/tasks/seq2seq.py (hunk 4 above)
Comment (Kelina) [reply jiant/tasks/seq2seq.py]: Changed. Thanks!

File: jiant/trainer.py (hunk 5 above)
Comment (sleepinyourhat) [reply jiant/trainer.py]: Yep!

File: jiant/tasks/__init__.py [hunk 8]
```diff
@@ -99,7 +99,7 @@
 ]
 
 # Seq2seq tasks
-ALL_SEQ2SEQ_TASKS = ["seg_wix"]
+ALL_SEQ2SEQ_TASKS = ["seg_wix", "wmt14_en_de"]
```
Comment (sleepinyourhat) [inline jiant/tasks/__init__.py]: Delete. (See above.)

File: jiant/tasks/seq2seq.py [hunk 9]
```diff
@@ -21,15 +22,19 @@
...
+@register_task("seg_wix", rel_path="seg/wix/", max_targ_v_size=200, valid_metric="accuracy")
+@register_task("wmt14_en_de", rel_path="wmt14/en_de/", max_targ_v_size=40000, valid_metric="BLEU")
 class Seq2SeqTask(SequenceGenerationTask):
     """Sequence-to-sequence Task"""
 
-    def __init__(self, path, max_seq_len, max_targ_v_size, name, **kw):
+    def __init__(self, path, max_seq_len, max_targ_v_size, name, valid_metric, **kw):
         super().__init__(name, **kw)
-        self.scorer2 = BooleanAccuracy()
+        if valid_metric == "accuracy":
```
Comment (sleepinyourhat) [inline jiant/tasks/seq2seq.py]: Make it clear in the header comment for this task class what the options are for this argument. (And generally spell out what all the arguments are.)

Comment (sleepinyourhat) [reply jiant/tasks/seq2seq.py]: (And what this task class is and isn't meant to support.)

File: jiant/tasks/seq2seq.py [hunk 10]
```diff
@@ -21,15 +22,19 @@
...
+            self.scorer2 = BooleanAccuracy()
+            self.val_metric = "%s_accuracy" % self.name
+        else:  # for MT
+            self.scorer2 = BLEU(exclude_indices=set([0, 1, 2, 3]))  # PAD, SOS, EOS, UNK
```
Comment (sleepinyourhat) [inline jiant/tasks/seq2seq.py]: How/where do we verify that this index order is right? Won't this break if we use something like RoBERTa (which we'll have to for our paper) that has its own vocab?

File: jiant/tasks/seq2seq.py [hunk 11]
```diff
@@ -71,7 +72,10 @@ def get_data_iter(self, path):
                 if len(row) < 2 or not row[0] or not row[1]:
                     continue
                 src_sent = tokenize_and_truncate(self._tokenizer_name, row[0], self.max_seq_len)
-                tgt_sent = tokenize_and_truncate(self._tokenizer_name, row[2], self.max_seq_len)
+                if self.val_metric == "%s_accuracy" % self.name:
+                    tgt_sent = tokenize_and_truncate(self._tokenizer_name, row[2], self.max_seq_len)
+                else:  # for MT
```
Comment (sleepinyourhat) [inline jiant/tasks/seq2seq.py]: This looks dangerous, especially since val_metric is user-settable. Find another way to distinguish the two task types.

File: jiant/tasks/seq2seq.py [hunk 12]
```diff
@@ -114,24 +118,52 @@ def _make_instance(input_, target):
...
+            val_metric = val_metric["BLEU"]
+        return {"perplexity": math.exp(avg_nll), metric_name: val_metric}
 
     def update_metrics(self, logits, labels, tagmask=None, predictions=None):
         # This doesn't require logits for now, since loss is updated in another part.
         assert logits is None and predictions is not None
 
-        if labels.shape[1] < predictions.shape[2]:
-            predictions = predictions[:, 0, : labels.shape[1]]
+        if self.val_metric != "%s_accuracy" % self.name:  # for MT
```
Comment (sleepinyourhat) [inline jiant/tasks/seq2seq.py]: See above.

File: jiant/tasks/seq2seq.py [hunk 13]
```diff
@@ -114,24 +118,52 @@ def _make_instance(input_, target):
...
+            else:
+                predictions = predictions[:, 0, :]
+                # Cut labels if predictions (without gold target) are shorter.
+                labels = labels[:, : predictions.shape[1]]
+                tagmask = tagmask[:, : predictions.shape[1]]
+            self.scorer2(predictions, labels, tagmask)
         return
 
     def get_prediction(self, voc_src, voc_trg, inputs, gold, output):
+        if self.val_metric == "%s_accuracy" % self.name:
```
Comment (sleepinyourhat) [inline jiant/tasks/seq2seq.py]: See above.

File: jiant/config/defaults.conf [hunk 14]
```diff
@@ -183,6 +183,8 @@ delete_checkpoints_when_done = 0  // If set, delete *all* saved checkpoints when
                                   // tests on any specific trained model. Will not apply if
                                   // keep_all_checkpoints is set.
 
+valid_metric = "BLEU"  // The validation metric used when a task has multiple options.
```
[the remark this answers is not in the document]
Comment (Kelina) [reply jiant/config/defaults.conf]: Yes, I think this should rather be a property of the task object. Thanks! 
Options are "BLEU" (to be changed to lowercase, see below) and "accuracy".

File: jiant/tasks/seq2seq.py (hunk 13 above)
Comment (Kelina) [reply jiant/tasks/seq2seq.py]: This part of the code for MT needs to receive the EOS index (related to the above comment regarding BLEU).

File: jiant/tasks/seq2seq.py (hunk 10 above)
Comment (Kelina) [reply jiant/tasks/seq2seq.py]: @HaokunLiu @pruksmhc Could you take a look at this?

File: jiant/tasks/seq2seq.py (hunk 11 above)
Comment (Kelina) [reply jiant/tasks/seq2seq.py]: @pruksmhc Could you please have a look at this, too? What would be the best way to distinguish between tasks? (I think Sam was against using task names.)

File: jiant/tasks/seq2seq.py (hunk 12 above)
Comment (Kelina) [reply jiant/tasks/seq2seq.py]: @pruksmhc

File: jiant/tasks/seq2seq.py (hunk 13 above)
Comment (Kelina) [reply jiant/tasks/seq2seq.py]: @pruksmhc

File: jiant/preprocess.py [hunk 15]
```diff
@@ -417,6 +417,7 @@ def _get_task(name, args, data_path, scratch_path):
             task_kw["probe_path"] = args["nli-prob"].probe_path
         if name in ALL_SEQ2SEQ_TASKS:
```
[the remark this answers is not in the document]
Comment (Kelina) [reply jiant/preprocess.py]: Will do - but just a quick note that I copied this from how it was done for other task (groups). Maybe this should be revised as well?

File: jiant/config/defaults.conf (hunk 14 above)
[the remark this answers is not in the document]
Comment (Kelina) [reply jiant/config/defaults.conf]: Done.

File: jiant/preprocess.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (Kelina) [reply jiant/preprocess.py]: Update: I don't think this can be done here a you suggested, since it's before the task object's creation. So will leave it as is for now.

File: jiant/tasks/__init__.py (hunk 8 above)
Comment (Kelina) [reply jiant/tasks/__init__.py]: Will stay for now (see above).

File: jiant/tasks/seq2seq.py (hunk 9 above)
Comment (Kelina) [reply jiant/tasks/seq2seq.py]: I added a comment that this comes from the config file. The options are listed there.

**Comment by sleepinyourhat** [comment]
Still work in progress, right?

**Comment by Kelina** [comment]
> Still pending from before:
> 
> * Could you say more about the exact comparison you're making for WMT. Test set vs test set? Dev set vs dev set? What's the exact target number you're trying to match, how close are you, and how does our setup differ?
> * If transformers (like RoBERTa) are supported, give an example config (or at least some instructions), if not, add an assert/error, and open an issue saying what we'd need to do to add support. We can't start our experiments without this.

I compared the development set results to reported (test set) results. We are slightly better, but our hyperparameters also differ slightly. From above: "I trained one model with similar hyperparameters to [https://arxiv.org/pdf/1508.04025.pdf](https://arxiv.org/pdf/1508.04025.pdf). The new model obtained 21.12 BLEU on the validation set, so it's roughly comparable."

The code runs with RoBERTa, but the results in my one-run experiment were worse (12.4 BLEU on dev); I am unsure if that's because of the hyperparameters. I added the config file I used for that.

**Comment by Kelina** [comment]
> Still work in progress, right?

The open remaining points are probably resolved better and faster by either @pruksmhc or @HaokunLiu. I tagged them above in the respective comments.

File: jiant/preprocess.py [hunk 16]
```diff
@@ -417,6 +417,7 @@ def _get_task(name, args, data_path, scratch_path):
...
             task_kw["max_targ_v_size"] = args.max_targ_word_v_size
+            task_kw["valid_metric"] = args.s2s.valid_metric
```
Comment (pruksmhc) [inline jiant/preprocess.py]: Why call this variable valid_metric as opposed to metric?

Comment (pruksmhc) [reply jiant/preprocess.py]: Ah validation metric. Let's rename to val_metric.

File: jiant/tasks/seq2seq.py [hunk 17]
```diff
@@ -21,15 +22,20 @@
...
 
-    def __init__(self, path, max_seq_len, max_targ_v_size, name, **kw):
+    def __init__(self, path, max_seq_len, max_targ_v_size, name, valid_metric, **kw):
         super().__init__(name, **kw)
-        self.scorer2 = BooleanAccuracy()
+        # Set scorer2 to compute the validation metric determined in the config file (see 's2s.valid_metric').
+        if valid_metric == "accuracy":
+            self.scorer2 = BooleanAccuracy()
+            self.val_metric = "%s_accuracy" % self.name
+        else:  # for MT
```
Comment (pruksmhc) [inline jiant/tasks/seq2seq.py]: Also why don't we set self.val_metric for BLEU case?

File: jiant/tasks/seq2seq.py (hunk 10 above)
Comment (pruksmhc) [reply jiant/tasks/seq2seq.py]: There might be a way to use the embedder https://github.com/nyu-mll/jiant/blob/9a47eea0f71021d8c714b4a455f6ad28793baa3c/jiant/pytorch_transformers_interface/modules.py and the _sep_id, _cls_id, _pad_id, and _unk_id variables.

File: jiant/tasks/seq2seq.py (hunk 11 above)
Comment (pruksmhc) [reply jiant/tasks/seq2seq.py]: Honestly, what I would do is create another class that inherits from Seq2SeqTask (MTTask for example), which uses BLEU. Then, the flow for everything in that class would be for BLEU, and avoids too many if-elses in the code.

File: jiant/tasks/seq2seq.py (hunk 12 above)
Comment (pruksmhc) [reply jiant/tasks/seq2seq.py]: Same comment

File: jiant/tasks/seq2seq.py (hunk 13 above)
Comment (pruksmhc) [reply jiant/tasks/seq2seq.py]: Same comment

File: scripts/wmt14/README [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+
```
Comment (pruksmhc) [inline scripts/wmt14/README]: Make sure you indicate in the jiant.info documentation that the documentation for MT is in scripts/wmt14/README.md

**Comment by pruksmhc** [comment]
@Kelina  where did you get the data?

**Comment by pruksmhc** [comment]
A number we should aim to _at_least_ have for RoBERTa is  27.14 BLEU (from https://openreview.net/pdf?id=Hyl7ygStwB),which is the number reported for using BERT to initialize the encoder of NMT).

**Comment by jeswan** [comment]
Are these changes still necessary? Planning to close all PRs to move jiant2 to this repo in the near future.
